vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+ *   number of iterations at current radius), and a new candidate will be chosen.
+ *
+ *
+ * @cite SimPL
+ * Original analytic placer with cut-spreading legalizing was intended for ASIC design, proposed in SimPL.
+ * SimPL: An Effective Placement Algorithm, Myung-Chul Kim, Dong-Jin Lee and Igor L. Markov
+ * http://www.ece.umich.edu/cse/awards/pdfs/iccad10-simpl.pdf
+ *
+ * @cite HeAP
+ * FPGA adaptation of SimPL, targeting FPGAs with heterogeneous blocks located at discrete locations.
+ * Analytical Placement for Heterogeneous FPGAs, Marcel Gort and Jason H. Anderson
+ * https://janders.eecg.utoronto.ca/pdfs/marcelfpl12.pdf
+ *
+ * @cite nextpnr
+ * An implementation of HeAP, which the cut-spreader and legalizer here is based off of. Implementation details
+ * have been modified for the architecture and netlist specification of VTR, and better performance.
+ * nextpnr -- Next Generation Place and Route,  placer_heap, David Shah <david@symbioticeda.com>
+ * https://github.com/YosysHQ/nextpnr
+ *
+ */
+#    include "vpr_context.h"
+
+// declaration of used types;
+class AnalyticPlacer;
+struct t_logical_block_type;
+
+// Cut-spreader, as described in HeAP/SimPL papers
+class CutSpreader {
+  public:
+    /*
+     * @brief: Constructor of CutSpreader
+     *
+     * @param analytic_placer	pointer to the analytic_placer that invokes this instance of CutSpreader
+     * 							passed for CutSpreader to directly access data members in analytic_placer such as
+     * 							blk_locs, blk_info, solve_blks, etc, without re-packaging the data to pass to
+     * 							CutSpreader.
+     *
+     * @param vpr_context		legal placement is returned by modifying vpr_context.mutable_placement(), which
+     * 							annealer then operates on. CutSpreader also gets netlist & device info from
+     * 							vpr_context.clustering().clb_nlist and vpr_context.device().
+     *
+     * @param blk_t				logical_block_type for CutSpreader to legalize. Currently can only legalize one
+     * 							type each time.
+     */
+    CutSpreader(AnalyticPlacer* analytic_placer, t_logical_block_type_ptr blk_t);
+
+    /*
+     * @brief: 	Executes the cut-spreader algorithm described in algorithm overview above.
+     * 			Does not include strict_legalize so placement result is not guaranteed to be legal.
+     * 			Strict_legalize must be run after for legal placement result, and for legal placement to
+     * 			be passed to annealer through vpr_ctx.
+     *
+     * 			Input placement is passed by data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to strict legalizer by modifying blk_locs in analytic_placer
+     */
+    void cutSpread();
+
+    /*
+     * @brief: 	Greedy strict legalize using algorithm described in algorithm overview above.
+     *
+     * 			Input illegal placement from data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to annealer by modifying vpr_ctx.mutable_placement()
```

Comment:
Don't you do a strict legalization in each solve/spread/legalize iteration?  (I thought that was the flow as analytic_placement.h comments).  If so, the result placement is presumably passed back through blk_locs, not vpr_ctx.mutable_placement().  Can you check this?
Explain which part of blk_locs is the input and which part is the output (there are multiple sub-members) for the cut spreader and the strict legalizer.